File: annotation-processor-common/src/main/java/ru/tinkoff/kora/annotation/processor/common/FieldFactory.java
package ru.tinkoff.kora.annotation.processor.common;

import com.palantir.javapoet.*;
import jakarta.annotation.Nullable;

import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class FieldFactory {
    private final Types types;
    private final Elements elements;
    @Nullable
    private final TypeSpec.Builder builder;
    private final MethodSpec.Builder constructor;
    private final Map<FieldKey, String> fields = new HashMap<>();
    private final String prefix;

    public String get(ClassName mapperType, TypeMirror mappedType, Element element) {
        var type = ParameterizedTypeName.get(mapperType, TypeName.get(mappedType).box());
        var tags = TagUtils.parseTagValue(element);
        var key = new FieldKey(type, tags);
        return this.fields.get(key);
    }

    public String get(ClassName mapperType, CommonUtils.MappingData mappingData, TypeMirror type) {
        var mapperClass = mappingData.mapperClass();
        if (mapperClass == null) {
            var mapperTypeName = ParameterizedTypeName.get(mapperType, TypeName.get(type).box());
            return this.fields.get(new FieldKey(mapperTypeName, mappingData.mapperTag()));
        }
        var mapperTypeName = TypeName.get(mapperClass);
        return this.fields.get(new FieldKey(mapperTypeName, mappingData.mapperTag()));
    }

    record FieldKey(TypeName typeName, @Nullable String tag) {}

    public FieldFactory(Types types, Elements elements, @Nullable TypeSpec.Builder builder, MethodSpec.Builder constructor, String prefix) {
        this.types = types;
        this.elements = elements;
        this.builder = builder;
        this.constructor = constructor;
        this.prefix = prefix;
    }

    public String add(TypeName typeName, @Nullable String tag) {
        var key = new FieldKey(typeName, tag);
        var existed = this.fields.get(key);
        if (existed != null) {
            return existed;
        }
        var name = this.prefix + (this.fields.size() + 1);
        this.fields.put(key, name);
        if (this.builder != null) {
            this.builder.addField(typeName, name, Modifier.PRIVATE, Modifier.FINAL);
            this.constructor.addStatement("this.$N = $N", name, name);
        }
        var parameter = ParameterSpec.builder(typeName, name);
        var tagAnnotation = TagUtils.makeAnnotationSpec(tag);
        if (tagAnnotation != null) {
            parameter.addAnnotation(tagAnnotation);
        }
        this.constructor.addParameter(parameter.build());
        return name;
    }

    public String add(TypeName typeName, CodeBlock initializer) {
        var key = new FieldKey(typeName, null);
        var existed = this.fields.get(key);
        if (existed != null) {
            return existed;
        }
        var name = this.prefix + (this.fields.size() + 1);
        this.fields.put(key, name);
        if (this.builder != null) {
            this.builder.addField(typeName, name, Modifier.PRIVATE, Modifier.FINAL);
            this.constructor.addStatement("this.$N = $L", name, initializer);
        } else {
            this.constructor.addStatement("var $N = $L", name, initializer);
        }
        return name;
    }

    public String add(TypeMirror typeMirror, @Nullable String tag) {
        var typeName = TypeName.get(typeMirror);
        var key = new FieldKey(typeName, tag);
        var existed = this.fields.get(key);
        if (existed != null) {
            return existed;
        }
        var name = this.prefix + (this.fields.size() + 1);
        this.fields.put(key, name);
        if (tag == null && CommonUtils.hasDefaultConstructorAndFinal(this.types, typeMirror)) {
            if (this.builder != null) {
                this.builder.addField(typeName, name, Modifier.PRIVATE, Modifier.FINAL);
                this.constructor.addStatement("this.$N = new $T()", name, typeName);
            } else {
                this.constructor.addStatement("var $N = new $T()", name, typeName);
            }
        } else {
            this.constructor.addParameter(typeName, name);
            if (this.builder != null) {
                this.builder.addField(typeName, name, Modifier.PRIVATE, Modifier.FINAL);
                this.constructor.addStatement("this.$N = $N", name, name);
            }
        }
        return name;
    }

    public String add(@Nullable CommonUtils.MappingData mapping, TypeName defaultType) {
        var tags = mapping == null
            ? null
            : mapping.mapperTag();
        var typeName = mapping == null
            ? defaultType
            : TypeName.get(Objects.requireNonNull(mapping.mapperClass()));
        var typeElement = typeName instanceof ParameterizedTypeName ptn
            ? this.elements.getTypeElement(ptn.rawType().canonicalName())
            : this.elements.getTypeElement(((ClassName) typeName).canonicalName());

        var key = new FieldKey(typeName, tags);
        var existed = this.fields.get(key);
        if (existed != null) {
            return existed;
        }
        var name = this.prefix + (this.fields.size() + 1);
        this.fields.put(key, name);
        if (tags == null && CommonUtils.hasDefaultConstructorAndFinal(typeElement)) {
            if (this.builder != null) {
                this.builder.addField(typeName, name, Modifier.PRIVATE, Modifier.FINAL);
                this.constructor.addStatement("this.$N = new $T()", name, typeName);
            } else {
                this.constructor.addStatement("var $N = new $T()", name, typeName);
            }
        } else {
            var parameter = ParameterSpec.builder(typeName, name);
            var tag = TagUtils.makeAnnotationSpec(tags);
            if (tag != null) {
                parameter.addAnnotation(tag);
            }
            this.constructor.addParameter(parameter.build());
            if (this.builder != null) {
                this.builder.addField(typeName, name, Modifier.PRIVATE, Modifier.FINAL);
                this.constructor.addStatement("this.$N = $N", name, name);
            }
        }
        return name;
    }
}

File: symbol-processor-common/src/main/kotlin/ru/tinkoff/kora/ksp/common/FieldFactory.kt
package ru.tinkoff.kora.ksp.common

import com.google.devtools.ksp.getConstructors
import com.google.devtools.ksp.isOpen
import com.google.devtools.ksp.symbol.KSAnnotated
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterSpec.Companion.builder
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.ksp.common.TagUtils.addTag

class FieldFactory(builder: TypeSpec.Builder, constructor: FunSpec.Builder, prefix: String) {
    private val builder: TypeSpec.Builder
    private val constructor: FunSpec.Builder
    private val fields: MutableMap<FieldKey, String> = HashMap()
    private val prefix: String

    operator fun get(mapperType: TypeName, resultMapperTag: String?): String? {
        return fields[FieldKey(mapperType, resultMapperTag)]
    }

    operator fun get(mapperType: ClassName, mappedType: KSType, element: KSAnnotated): String {
        val type = mapperType.parameterizedBy(mappedType.toTypeName().copy(false))
        val tags = TagUtils.parseTagValue(element)
        val key = FieldKey(type, tags)
        return fields[key]!!
    }

    operator fun get(mapperClass: KSType, resultMapperTag: String?): String {
        val mapperType: TypeName = mapperClass.toTypeName()
        return fields[FieldKey(mapperType, resultMapperTag)]!!
    }

    internal data class FieldKey(val typeName: TypeName, val tag: String?)

    init {
        this.builder = builder
        this.constructor = constructor
        this.prefix = prefix
    }

    fun add(typeName: TypeName, tag: String?): String {
        val key = FieldKey(typeName, tag)
        val existed = fields[key]
        if (existed != null) {
            return existed
        }
        val name = prefix + (fields.size + 1)
        fields[key] = name
        builder.addProperty(name, typeName, KModifier.PRIVATE)
        val parameter = builder(name, typeName)
        parameter.addTag(tag)
        constructor.addParameter(parameter.build())
        constructor.addStatement("this.%N = %N", name, name)
        return name
    }

    fun add(typeName: TypeName, initializer: CodeBlock): String {
        val key = FieldKey(typeName, null)
        val existed = fields[key]
        if (existed != null) {
            return existed
        }
        val name = prefix + (fields.size + 1)
        fields[key] = name
        builder.addProperty(name, typeName, KModifier.PRIVATE)
        constructor.addStatement("this.%N = %L", name, initializer)
        return name
    }

    fun add(typeMirror: KSType, tag: String?): String {
        val typeName = typeMirror.toTypeName()
        val key = FieldKey(typeName, tag)
        val existed = fields[key]
        if (existed != null) {
            return existed
        }
        val name = prefix + (fields.size + 1)
        fields[key] = name
        builder.addProperty(name, typeName, KModifier.PRIVATE)
        val decl = typeMirror.declaration
        if (tag == null && decl is KSClassDeclaration && !decl.isOpen() && decl.getConstructors().count() == 1 && decl.getConstructors().first().parameters.isEmpty()) {
            constructor.addStatement("this.%N = %T()", name, typeName)
        } else {
            constructor.addParameter(name, typeName)
            constructor.addStatement("this.%N = %N", name, name)
        }
        return name
    }

    fun add(mapping: MappingData?, fieldParserType: TypeName): String {
        val typeName = mapping?.mapper?.toTypeName() ?: fieldParserType
        val tag = mapping?.tag
        val key = FieldKey(typeName, tag)
        val existed = fields[key]
        if (existed != null) {
            return existed
        }
        val name = prefix + (fields.size + 1)
        fields[key] = name
        builder.addProperty(name, typeName, KModifier.PRIVATE)
        val decl = mapping?.mapper?.declaration
        if (tag == null && decl is KSClassDeclaration && !decl.isOpen() && decl.getConstructors().count() == 1 && decl.getConstructors().first().parameters.isEmpty()) {
            constructor.addStatement("this.%N = %T()", name, typeName)
        } else {
            val parameter = builder(name, typeName)
                .addTag(tag)
            constructor.addParameter(parameter.build())
            constructor.addStatement("this.%N = %N", name, name)
        }
        return name
    }
}
